File: SinglyFariz.java
// CLASS NODE MANUAL
class Node {
    String nim;
    String nama;
    String jk;
    Node next;

    public Node(String nim, String nama, String jk) {
        this.nim = nim;
        this.nama = nama;
        this.jk = jk;
        this.next = null;
    }
}

public class SinglyFariz {
    public static void main(String[] args) {
        Node head = null;
        int count = 0;
        int choice;

        java.util.Scanner sc = new java.util.Scanner(System.in);

        // HEADER
        System.out.println("\n=================================================");
        System.out.println("       TUGAS STRUKTUR DATA LINKED LIST           ");
        System.out.println("=================================================");
        System.out.println(" Nama    : Fariz Husain Albar");
        System.out.println(" Kelas   : Informatika A");
        System.out.println(" NIM     : 24106050011");
        System.out.println(" Matkul  : Struktur Data");
        System.out.println("=================================================");
        System.out.println(" >> Tekan ENTER untuk masuk ke Menu Program...");
        sc.nextLine();

        while (true) {
            System.out.println("\n=== MENU MAHASISWA (Count: " + count + ") ===");
            System.out.println("1. Insert at beginning");
            System.out.println("2. Insert at given position");
            System.out.println("3. Insert at end");
            System.out.println("4. Delete from beginning");
            System.out.println("5. Delete given position");
            System.out.println("6. Delete from end");
            System.out.println("7. Delete first occurence (by NIM)");
            System.out.println("8. Show data");
            System.out.println("9. Exit");
            System.out.print("Pilih menu: ");

            if (sc.hasNextInt()) {
                choice = sc.nextInt();
                sc.nextLine();
            } else {
                System.out.println(">> Input salah! Harap masukkan angka.");
                sc.next();
                continue;
            }

            // 1. INSERT AT BEGINNING
            if (choice == 1) {
                // Validasi NIM
                String nim = "";
                while (true) {
                    System.out.print("Masukkan NIM: ");
                    nim = sc.nextLine();
                    if (!nim.matches("[0-9]+")) {
                        System.out.println(">> ERROR: NIM harus angka!");
                        continue;
                    }
                    boolean exists = false;
                    Node temp = head;
                    while (temp != null) {
                        if (temp.nim.equals(nim)) { exists = true; break; }
                        temp = temp.next;
                    }
                    if (exists) System.out.println(">> ERROR: NIM sudah ada!");
                    else break;
                }

                // Validasi Nama
                String nama = "";
                while (true) {
                    System.out.print("Masukkan Nama: ");
                    nama = sc.nextLine();
                    boolean exists = false;
                    Node temp = head;
                    while (temp != null) {
                        if (temp.nama.equalsIgnoreCase(nama)) { exists = true; break; }
                        temp = temp.next;
                    }
                    if (exists) System.out.println(">> ERROR: Nama sudah ada!");
                    else break;
                }

                // Validasi JK
                String jk = "";
                while (true) {
                    System.out.print("Masukkan Jenis Kelamin (L/P): ");
                    String inputJk = sc.nextLine().toUpperCase();
                    if (inputJk.equals("L") || inputJk.equals("P")) {
                        jk = inputJk;
                        break;
                    }
                    System.out.println(">> ERROR: Input salah (L/P)!");
                }

                Node newNode = new Node(nim, nama, jk);
                if (head == null) {
                    head = newNode;
                } else {
                    newNode.next = head;
                    head = newNode;
                }
                count++;
                System.out.println(">> SUKSES: Data insert di depan.");
            }

            // 2. INSERT AT GIVEN POSITION
            else if (choice == 2) {
                int pos = 0;
                while (true) {
                    System.out.print("Masukkan Posisi Insert (1 - " + (count + 1) + "): ");
                    if (sc.hasNextInt()) {
                        pos = sc.nextInt();
                        sc.nextLine();
                        if (pos >= 1 && pos <= count + 1) break;
                        System.out.println(">> ERROR: Posisi tidak valid!");
                    } else {
                        System.out.println(">> ERROR: Harap masukkan angka!");
                        sc.next();
                    }
                }

                // INPUT DATA SAMA SEPERTI DI ATAS
                String nim = "";
                while (true) {
                    System.out.print("Masukkan NIM: "); nim = sc.nextLine();
                    if (!nim.matches("[0-9]+")) { System.out.println(">> ERROR: NIM harus angka!"); continue; }
                    boolean exists = false; Node temp = head;
                    while (temp != null) { if (temp.nim.equals(nim)) { exists = true; break; } temp = temp.next; }
                    if (exists) System.out.println(">> ERROR: NIM sudah ada!"); else break;
                }
                String nama = "";
                while (true) {
                    System.out.print("Masukkan Nama: "); nama = sc.nextLine();
                    boolean exists = false; Node temp = head;
                    while (temp != null) { if (temp.nama.equalsIgnoreCase(nama)) { exists = true; break; } temp = temp.next; }
                    if (exists) System.out.println(">> ERROR: Nama sudah ada!"); else break;
                }
                String jk = "";
                while (true) {
                    System.out.print("Masukkan Jenis Kelamin (L/P): "); String inputJk = sc.nextLine().toUpperCase();
                    if (inputJk.equals("L") || inputJk.equals("P")) { jk = inputJk; break; }
                    System.out.println(">> ERROR: Input salah (L/P)!");
                }

                Node newNode = new Node(nim, nama, jk);
                if (pos == 1) {
                    newNode.next = head;
                    head = newNode;
                } else {
                    Node current = head;
                    for (int i = 1; i < pos - 1; i++) {
                        current = current.next;
                    }
                    newNode.next = current.next;
                    current.next = newNode;
                }
                count++;
                System.out.println(">> SUKSES: Data insert di posisi " + pos);
            }

            // 3. INSERT AT END
            else if (choice == 3) {
                String nim = "";
                while (true) {
                    System.out.print("Masukkan NIM: "); nim = sc.nextLine();
                    if (!nim.matches("[0-9]+")) { System.out.println(">> ERROR: NIM harus angka!"); continue; }
                    boolean exists = false; Node temp = head;
                    while (temp != null) { if (temp.nim.equals(nim)) { exists = true; break; } temp = temp.next; }
                    if (exists) System.out.println(">> ERROR: NIM sudah ada!"); else break;
                }
                String nama = "";
                while (true) {
                    System.out.print("Masukkan Nama: "); nama = sc.nextLine();
                    boolean exists = false; Node temp = head;
                    while (temp != null) { if (temp.nama.equalsIgnoreCase(nama)) { exists = true; break; } temp = temp.next; }
                    if (exists) System.out.println(">> ERROR: Nama sudah ada!"); else break;
                }
                String jk = "";
                while (true) {
                    System.out.print("Masukkan Jenis Kelamin (L/P): "); String inputJk = sc.nextLine().toUpperCase();
                    if (inputJk.equals("L") || inputJk.equals("P")) { jk = inputJk; break; }
                    System.out.println(">> ERROR: Input salah (L/P)!");
                }

                Node newNode = new Node(nim, nama, jk);
                if (head == null) {
                    head = newNode;
                } else {
                    Node temp = head;
                    while (temp.next != null) {
                        temp = temp.next;
                    }
                    temp.next = newNode;
                }
                count++;
                System.out.println(">> SUKSES: Data insert di akhir.");
            }

            // 4. DELETE FROM BEGINNING
            else if (choice == 4) {
                if (head == null) {
                    System.out.println(">> GAGAL: LinkedList Kosong!");
                } else {
                    head = head.next;
                    count--;
                    System.out.println(">> SUKSES: Data pertama dihapus.");
                }
            }

            // 5. DELETE GIVEN POSITION
            else if (choice == 5) {
                if (head == null) {
                    System.out.println(">> GAGAL: LinkedList Kosong!");
                } else {
                    int pos = 0;
                    while (true) {
                        System.out.print("Masukkan Posisi Hapus (1 - " + count + "): ");
                        if (sc.hasNextInt()) {
                            pos = sc.nextInt(); sc.nextLine();
                            if (pos >= 1 && pos <= count) break;
                            System.out.println(">> ERROR: Posisi tidak valid!");
                        } else { System.out.println(">> ERROR: Harap masukkan angka!"); sc.next(); }
                    }

                    if (pos == 1) {
                        head = head.next;
                    } else {
                        Node temp = head;
                        for (int i = 1; i < pos - 1; i++) {
                            temp = temp.next;
                        }
                        Node nodeToDelete = temp.next;
                        temp.next = nodeToDelete.next;
                    }
                    count--;
                    System.out.println(">> SUKSES: Data posisi " + pos + " dihapus.");
                }
            }

            // 6. DELETE FROM END
            else if (choice == 6) {
                if (head == null) {
                    System.out.println(">> GAGAL: LinkedList Kosong!");
                } else if (head.next == null) {
                    head = null;
                    count--;
                    System.out.println(">> SUKSES: Data terakhir dihapus.");
                } else {
                    Node temp = head;
                    while (temp.next.next != null) {
                        temp = temp.next;
                    }
                    temp.next = null;
                    count--;
                    System.out.println(">> SUKSES: Data terakhir dihapus.");
                }
            }

            // 7. DELETE FIRST OCCURRENCE (BY NIM)
            else if (choice == 7) {
                if (head == null) {
                    System.out.println(">> GAGAL: LinkedList Kosong!");
                } else {
                    System.out.print("Masukkan NIM yang akan dihapus: ");
                    String targetNim = sc.nextLine();

                    if (head.nim.equals(targetNim)) {
                        head = head.next;
                        count--;
                        System.out.println(">> SUKSES: NIM " + targetNim + " dihapus.");
                    } else {
                        Node temp = head;
                        boolean found = false;
                        while (temp.next != null) {
                            if (temp.next.nim.equals(targetNim)) {
                                temp.next = temp.next.next;
                                count--;
                                found = true;
                                break;
                            }
                            temp = temp.next;
                        }
                        if (found) System.out.println(">> SUKSES: NIM " + targetNim + " dihapus.");
                        else System.out.println(">> GAGAL: NIM tidak ditemukan.");
                    }
                }
            }

            // 8. SHOW DATA
            else if (choice == 8) {
                System.out.println("\nData Mahasiswa (LinkedList):");
                System.out.println("+-----+--------------+---------------------------+-----+");
                System.out.printf("| %-3s | %-12s | %-25s | %-3s |%n", "No", "NIM", "Nama", "L/P");
                System.out.println("+-----+--------------+---------------------------+-----+");

                if (head == null) {
                    System.out.printf("| %-50s |%n", "                  DATA KOSONG");
                } else {
                    Node temp = head;
                    int i = 1;
                    while (temp != null) {
                        System.out.printf("| %-3d | %-12s | %-25s | %-3s |%n",
                                i, temp.nim, temp.nama, temp.jk);
                        temp = temp.next;
                        i++;
                    }
                }
                System.out.println("+-----+--------------+---------------------------+-----+");
            }

            // 9. EXIT
            else if (choice == 9) {
                System.out.println("Keluar program...");
                break;
            } else {
                System.out.println(">> Menu tidak valid!");
            }
        }
        sc.close();
    }
}

File: SinglyFariz.py
# CLASS NODE
# Pengganti struct/class di Java
class Node:
    def __init__(self, nim, nama, jk):
        self.nim = nim
        self.nama = nama
        self.jk = jk
        self.next = None  # Pointer ke node berikutnya

# MAIN PROGRAM
if __name__ == "__main__":
    head = None  # Kepala rantai (awalnya kosong)
    count = 0    # Penghitung jumlah data

    # HEADER
    print("\n" + "="*49)
    print("       TUGAS STRUKTUR DATA LINKED LIST           ")
    print("="*49)
    print(" Nama    : Fariz Husain Albar")
    print(" Kelas   : Informatika A")
    print(" NIM     : 24106050011")
    print(" Matkul  : Struktur Data")
    print("="*49)
    input(" >> Tekan ENTER untuk masuk ke Menu Program...")

    while True:
        print(f"\n=== MENU MAHASISWA (Count: {count}) ===")
        print("1. Insert at beginning")
        print("2. Insert at given position")
        print("3. Insert at end")
        print("4. Delete from beginning")
        print("5. Delete given position")
        print("6. Delete from end")
        print("7. Delete first occurence (by NIM)")
        print("8. Show data")
        print("9. Exit")

        try:
            choice_input = input("Pilih menu: ")
            if not choice_input.isdigit():
                print(">> Input salah! Harap masukkan angka.")
                continue
            choice = int(choice_input)
        except ValueError:
            print(">> Input salah! Harap masukkan angka.")
            continue

        # LOGIKA MENU

        # 1. INSERT AT BEGINNING
        if choice == 1:
            # Validasi NIM
            nim = ""
            while True:
                nim = input("Masukkan NIM: ")
                if not nim.isdigit():
                    print(">> ERROR: NIM harus angka!")
                    continue

                # Cek Duplikasi
                exists = False
                temp = head
                while temp is not None:
                    if temp.nim == nim:
                        exists = True
                        break
                    temp = temp.next

                if exists:
                    print(">> ERROR: NIM sudah ada!")
                else:
                    break

            # Validasi Nama
            nama = ""
            while True:
                nama = input("Masukkan Nama: ")
                exists = False
                temp = head
                while temp is not None:
                    if temp.nama.lower() == nama.lower():
                        exists = True
                        break
                    temp = temp.next

                if exists:
                    print(">> ERROR: Nama sudah ada!")
                else:
                    break

            # Validasi JK
            jk = ""
            while True:
                jk = input("Masukkan Jenis Kelamin (L/P): ").upper()
                if jk == "L" or jk == "P":
                    break
                print(">> ERROR: Input salah (L/P)!")

            # PROSES INSERT DEPAN
            new_node = Node(nim, nama, jk)
            if head is None:
                head = new_node
            else:
                new_node.next = head
                head = new_node

            count += 1
            print(">> SUKSES: Data insert di depan.")

        # 2. INSERT AT GIVEN POSITION
        elif choice == 2:
            pos = 0
            while True:
                try:
                    pos_input = input(f"Masukkan Posisi Insert (1 - {count + 1}): ")
                    if not pos_input.isdigit():
                        print(">> ERROR: Harap masukkan angka!")
                        continue
                    pos = int(pos_input)
                    if 1 <= pos <= count + 1:
                        break
                    print(">> ERROR: Posisi tidak valid!")
                except ValueError:
                    print(">> ERROR: Harap masukkan angka!")

            # INPUT DATA SAMA SEPERTI DI ATAS
            nim = ""
            while True:
                nim = input("Masukkan NIM: ")
                if not nim.isdigit(): print(">> ERROR: NIM harus angka!"); continue
                exists = False; temp = head
                while temp:
                    if temp.nim == nim: exists = True; break
                    temp = temp.next
                if exists: print(">> ERROR: NIM sudah ada!");
                else: break

            nama = ""
            while True:
                nama = input("Masukkan Nama: ")
                exists = False; temp = head
                while temp:
                    if temp.nama.lower() == nama.lower(): exists = True; break
                    temp = temp.next
                if exists: print(">> ERROR: Nama sudah ada!");
                else: break

            jk = ""
            while True:
                jk = input("Masukkan Jenis Kelamin (L/P): ").upper()
                if jk in ["L", "P"]: break
                print(">> ERROR: Input salah (L/P)!")

            # PROSES INSERT POSISI
            new_node = Node(nim, nama, jk)
            if pos == 1:
                new_node.next = head
                head = new_node
            else:
                current = head
                # Loop sampai node sebelum posisi target (pos-2 karena index 0)
                for _ in range(1, pos - 1):
                    current = current.next

                new_node.next = current.next
                current.next = new_node

            count += 1
            print(f">> SUKSES: Data insert di posisi {pos}")

        # 3. INSERT AT END
        elif choice == 3:
            # INPUT DATA SAMA
            nim = ""
            while True:
                nim = input("Masukkan NIM: ")
                if not nim.isdigit(): print(">> ERROR: NIM harus angka!"); continue
                exists = False; temp = head
                while temp:
                    if temp.nim == nim: exists = True; break
                    temp = temp.next
                if exists: print(">> ERROR: NIM sudah ada!");
                else: break

            nama = ""
            while True:
                nama = input("Masukkan Nama: ")
                exists = False; temp = head
                while temp:
                    if temp.nama.lower() == nama.lower(): exists = True; break
                    temp = temp.next
                if exists: print(">> ERROR: Nama sudah ada!");
                else: break

            jk = ""
            while True:
                jk = input("Masukkan Jenis Kelamin (L/P): ").upper()
                if jk in ["L", "P"]: break
                print(">> ERROR: Input salah (L/P)!")

            # PROSES INSERT BELAKANG
            new_node = Node(nim, nama, jk)
            if head is None:
                head = new_node
            else:
                temp = head
                while temp.next is not None:
                    temp = temp.next
                temp.next = new_node

            count += 1
            print(">> SUKSES: Data insert di akhir.")

        # 4. DELETE FROM BEGINNING
        elif choice == 4:
            if head is None:
                print(">> GAGAL: LinkedList Kosong!")
            else:
                head = head.next
                count -= 1
                print(">> SUKSES: Data pertama dihapus.")

        # 5. DELETE GIVEN POSITION
        elif choice == 5:
            if head is None:
                print(">> GAGAL: LinkedList Kosong!")
            else:
                pos = 0
                while True:
                    try:
                        pos_input = input(f"Masukkan Posisi Hapus (1 - {count}): ")
                        if not pos_input.isdigit():
                            print(">> ERROR: Harap masukkan angka!")
                            continue
                        pos = int(pos_input)
                        if 1 <= pos <= count: break
                        print(">> ERROR: Posisi tidak valid!")
                    except ValueError:
                        print(">> ERROR: Harap masukkan angka!")

                if pos == 1:
                    head = head.next
                else:
                    temp = head
                    for _ in range(1, pos - 1):
                        temp = temp.next

                    # Skip node yang dihapus
                    node_to_delete = temp.next
                    temp.next = node_to_delete.next

                count -= 1
                print(f">> SUKSES: Data posisi {pos} dihapus.")

        # 6. DELETE FROM END
        elif choice == 6:
            if head is None:
                print(">> GAGAL: LinkedList Kosong!")
            elif head.next is None:
                head = None
                count -= 1
                print(">> SUKSES: Data terakhir dihapus.")
            else:
                temp = head
                while temp.next.next is not None:
                    temp = temp.next

                temp.next = None
                count -= 1
                print(">> SUKSES: Data terakhir dihapus.")

        # 7. DELETE FIRST OCCURRENCE (BY NIM)
        elif choice == 7:
            if head is None:
                print(">> GAGAL: LinkedList Kosong!")
            else:
                target_nim = input("Masukkan NIM yang akan dihapus: ")

                if head.nim == target_nim:
                    head = head.next
                    count -= 1
                    print(f">> SUKSES: NIM {target_nim} dihapus.")
                else:
                    temp = head
                    found = False
                    while temp.next is not None:
                        if temp.next.nim == target_nim:
                            temp.next = temp.next.next
                            count -= 1
                            found = True
                            break
                        temp = temp.next

                    if found:
                        print(f">> SUKSES: NIM {target_nim} dihapus.")
                    else:
                        print(">> GAGAL: NIM tidak ditemukan.")

        # 8. SHOW DATA
        elif choice == 8:
            print("\nData Mahasiswa (LinkedList):")
            print("+-----+--------------+---------------------------+-----+")
            print(f"| {'No':<3} | {'NIM':<12} | {'Nama':<25} | {'L/P':<3} |")
            print("+-----+--------------+---------------------------+-----+")

            if head is None:
                print(f"| {'DATA KOSONG':^50} |")
            else:
                temp = head
                i = 1
                while temp is not None:
                    print(f"| {i:<3} | {temp.nim:<12} | {temp.nama:<25} | {temp.jk:<3} |")
                    temp = temp.next
                    i += 1
            print("+-----+--------------+---------------------------+-----+")

        # 9. EXIT
        elif choice == 9:
            print("Keluar program...")
            break

        else:
            print(">> Menu tidak valid!")